libp2p/module.go

```diff
@@ -0,0 +1,367 @@
+	// Topic is used to `#Publish` messages.
+	mod.topic, err = mod.pubsub.Join(protocol.DefaultTopicStr)
+	if err != nil {
+		return fmt.Errorf("unable to join pubsub topic: %w", err)
+	}
+
+	// Subscription is notified when a new message is received on the topic.
+	mod.subscription, err = mod.topic.Subscribe()
+	if err != nil {
+		return fmt.Errorf("subscribing to pubsub topic: %w", err)
+	}
+
+	mod.network, err = network.NewLibp2pNetwork(mod.GetBus(), mod.logger, mod.host, mod.topic)
+	if err != nil {
+		return fmt.Errorf("creating network: %w", err)
+	}
+
+	// Don't handle streams or read from the subscription in client debug mode.
+	if !mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
+		mod.host.SetStreamHandler(protocol.PoktProtocolID, mod.handleStream)
+		go mod.readFromSubscription(ctx)
+	}
+	return nil
+}
+
+func (mod *libp2pModule) Stop() error {
+	return mod.host.Close()
+}
+
+func (mod *libp2pModule) Broadcast(msg *anypb.Any) error {
+	c := &messaging.PocketEnvelope{
+		Content: msg,
+	}
+	data, err := proto.MarshalOptions{Deterministic: true}.Marshal(c)
+	if err != nil {
+		return err
+	}
+	mod.logger.Info().Msg("broadcasting message to network")
+
+	return mod.network.NetworkBroadcast(data)
+}
+
+func (mod *libp2pModule) Send(addr crypto.Address, msg *anypb.Any) error {
+	c := &messaging.PocketEnvelope{
+		Content: msg,
+	}
+	data, err := proto.MarshalOptions{Deterministic: true}.Marshal(c)
+	if err != nil {
+		return err
+	}
+
+	return mod.network.NetworkSend(data, addr)
+}
+
+func (mod *libp2pModule) GetAddress() (crypto.Address, error) {
+	privateKey, err := crypto.NewPrivateKey(mod.cfg.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return privateKey.Address(), nil
+}
+
+// handleStream is called each time a peer establishes a new stream with this
+// module's libp2p `host.Host`.
+func (mod *libp2pModule) handleStream(stream libp2pNetwork.Stream) {
+	peer, err := network.PeerFromLibp2pStream(stream)
+	if err != nil {
+		mod.logger.Error().Err(err).
+			Str("address", peer.Address.String()).
+			Msg("parsing remote peer public key")
+
+		if err = stream.Close(); err != nil {
+			mod.logger.Error().Err(err)
+		}
+	}
+
+	if err := mod.network.AddPeerToAddrBook(peer); err != nil {
+		mod.logger.Error().Err(err).
+			Str("address", peer.Address.String()).
+			Msg("adding remote peer to address book")
+	}
+
+	go mod.readStream(stream)
+}
+
+// readStream is intended to be called in a goroutine. It continuously reads from
+// the given stream for handling at the network level. Used for handling "direct"
+// messages (i.e. one specific target node).
+func (mod *libp2pModule) readStream(stream libp2pNetwork.Stream) {
+	closeStream := func() {
+		if err := stream.Close(); err != nil {
+			mod.logger.Error().Err(err)
+		}
+	}
+
+	// NB: time out if no data is sent to free resources.
+	if err := stream.SetReadDeadline(newReadStreamDeadline()); err != nil {
+		mod.logger.Error().Err(err).Msg("setting stream read deadline")
```

Comment:
It seems that we are settling on this pattern (Added the Bool(true)) for this kind of thing to help with 🪵 🔍
```suggestion
		mod.logger.Error().Err(err).
        Bool("TODO", true).
        Msg("setting stream read deadline")
```

Reply:
@deblasis can you elaborate on  (or link to) the thinking behind this?

Reply:
Of course. 

We have a bunch of places in our codebase where we log a message and we also attach a `TODO`: true

Example:
![image](https://user-images.githubusercontent.com/29378614/222149797-9f08adc8-ee19-48a1-ae98-fb1e0eb2aaaa.png)

As we mature our observability stack, it could be useful to search for these and also to inform the users that something is still in development and the log entries are merely placeholders for future real functionality

Reply:
We also have #478